src/quotamodel.c

```diff
@@ -900,6 +897,75 @@ merge_uncommitted_table_to_oidlist(List *oidlist)
 	return oidlist;
 }
 
+static TableSizeEntry *
+get_tsentry(Oid oid, int16 segid)
+{
+	bool              found;
+	TableSizeEntryKey key     = {.reloid = oid, .id = TableSizeEntryId(segid)};
+	HASHACTION        action  = check_hash_fullness(table_size_map, MAX_NUM_TABLE_SIZE_ENTRIES, table_size_map_warning,
+	                                                table_size_map_last_overflow_report);
+	TableSizeEntry   *tsentry = hash_search(table_size_map, &key, action, &found);
+
+	if (!found && tsentry != NULL)
+	{
+		memset(tsentry->totalsize, 0, sizeof(tsentry->totalsize));
+		tsentry->owneroid      = InvalidOid;
+		tsentry->namespaceoid  = InvalidOid;
+		tsentry->tablespaceoid = InvalidOid;
+		tsentry->flag          = 0;
+
+		int seg_st = TableSizeEntrySegidStart(tsentry);
+		int seg_ed = TableSizeEntrySegidEnd(tsentry);
+		for (int j = seg_st; j < seg_ed; j++) TableSizeEntrySetFlushFlag(tsentry, j);
+	}
+
+	/* mark tsentry is_exist */
+	if (tsentry) set_table_size_entry_flag(tsentry, TABLE_EXIST);
+
+	return tsentry;
+}
+
+void
+update_active_table_size(Oid oid, int64 size, int16 segid)
+{
+	TableSizeEntry *tsentry = get_tsentry(oid, segid);
+
+	if (tsentry == NULL)
+	{
+		/* Too many tables have been added to the table_size_map, to avoid diskquota using
+		   too much share memory, just return. The diskquota won't work correctly
+		   anymore. */
+		return;
+	}
+
+	if (segid == -1)
+	{
+		/* pretend process as utility mode, and append the table size on master */
+		Gp_role = GP_ROLE_UTILITY;
+
+		/* when segid is -1, the size is the sum of size of master and all segments */
+		size += calculate_table_size(oid);
+
+		Gp_role = GP_ROLE_DISPATCH;
```

Comment:
wouldn't it be better to assign previously saved value of Gp_role here instead of constant? Perhaps with assert for Gp_role == GP_ROLE_DISPATCH.

Reply:
This has nothing to do with the current patch.

Reply:
Please explain.

Reply:
This was also the case before the patch.

Reply:
I'm afraid that in this case this is not possible.

Reply:
How is this refactoring with extraction of update_active_table_size function related to initial goal "Refactor diskquota local hashmap with active tables"? How is it related to the objective to save memory at init?

Reply:
The contents of the `update_active_table_size` function before the patch was used to update the size of active tables in the `calculate_table_disk_usage` function. But now the `calculate_table_disk_usage` function knows nothing about active tables, so I moved the size update to a separate function.

Reply:
Pls explain in more details the change in calculate_table_disk_usage() logic. Previously it recalculated table sizes only for the tables not in active list. Does it recalculates sizes for all tables now?

Reply:
> Previously it recalculated table sizes only for the tables not in active list.

Really?
https://github.com/arenadata/diskquota/blob/0d2375a0b4990eea06b622958f0f9538dc435b95/src/quotamodel.c#L1041-L1073

Reply:
So you want to say that this logic has not changed, right?

Reply:
Yes, [except](https://github.com/arenadata/diskquota/pull/46#discussion_r1897868910)

Reply:
I suggest to either describe this active tables scan/detection logic change in the PR description, and explain why it is good, where we can benefit from it and why we don't test is specifically.
Or, alternatively, revert this logic change.

P.S. are there tests which specifically cover this refactoring?